http.go

```diff
@@ -1936,24 +1921,27 @@ func readBody(r *bufio.Reader, contentLength int, maxBodySize int, dst []byte) (
 var errChunkedStream = errors.New("chunked stream")
 
 func readBodyWithStreaming(r *bufio.Reader, contentLength int, maxBodySize int, dst []byte) (b []byte, err error) {
+	if contentLength == -1 {
+		// handled in requestStream.Read()
+		return b, errChunkedStream
+	}
+
 	dst = dst[:0]
-	switch {
-	case contentLength >= 0 && maxBodySize >= contentLength:
-		readN := maxBodySize
-		if contentLength > 8*1024 {
-			readN = 8 * 1024
-		}
+
+	readN := maxBodySize
+	if readN > contentLength {
+		readN = contentLength
```

Comment:
Without this we would wait forever on bodies smaller than 8kb as we would block on reading from a connection that never had any more data.